service-proxy.js

```diff
@@ -514,38 +520,71 @@ function refreshServicePeerPartially(serviceName, hostPort) {
     );
     var stop = (start + length) % workers.length;
 
-    self.logger.info('Refreshing service peer affinity', self.extendLogInfo({
-        serviceName: serviceName,
-        serviceHostPort: hostPort,
+    return {
+        start: start,
+        stop: stop,
         relayIndex: relayIndex,
         relays: relays,
         workers: workers,
-        start: start,
-        stop: stop,
         length: length
+    };
+};
+
+ServiceDispatchHandler.prototype.refreshServicePeerPartially =
+function refreshServicePeerPartially(serviceName, hostPort) {
+    var self = this;
+
+    var range = self.computePartialRange(serviceName, hostPort);
+
+    self.logger.info('Refreshing service peer affinity', self.extendLogInfo({
+        serviceName: serviceName,
+        serviceHostPort: hostPort,
+        relayIndex: range.relayIndex,
+        relays: range.relays,
+        workers: range.workers,
+        start: range.start,
+        stop: range.stop,
+        length: range.length
     }));
 
     // Open connections to affine peers
     var index;
     var peer;
-    if (start <= stop) { // ... start WITHIN stop ...
-        for (index = start; index < stop; index++) {
-            peer = self.getServicePeer(serviceName, workers[index]);
-            peer.connectTo();
+
+    // fully connected
+    if (range.start === range.stop) {
+        for (index = 0; index < range.workers.length; index++) {
+            peer = self.getServicePeer(serviceName, range.workers[index]);
+            if (!peer.isConnected('out')) {
+                peer.connectTo();
+            }
+        }
+    } else if (range.start <= range.stop) { // ... start WITHIN stop ...
+        for (index = range.start; index < range.stop; index++) {
+            peer = self.getServicePeer(serviceName, range.workers[index]);
+            if (!peer.isConnected('out')) {
+                peer.connectTo();
+            }
         }
     } else { // BEFORE stop ... start AFTER
-        for (index = start; index < workers.length; index++) {
-            peer = self.getServicePeer(serviceName, workers[index]);
-            peer.connectTo();
+        for (index = 0; index < range.stop; index++) {
+            peer = self.getServicePeer(serviceName, range.workers[index]);
+            if (!peer.isConnected('out')) {
+                peer.connectTo();
+            }
         }
-        for (index = 0; index < stop; index++) {
-            peer = self.getServicePeer(serviceName, workers[index]);
-            peer.connectTo();
+        for (index = range.start; index < range.workers.length; index++) {
+            peer = self.getServicePeer(serviceName, range.workers[index]);
+            if (!peer.isConnected('out')) {
+                peer.connectTo();
+            }
```

Comment:
This will probably work well enough, but I don’t think it’ll properly reap connections that are still advertising but no longer affine. We may need to follow up by decoupling peer membership (retained by ad) from affine membership (connected by ad if affine, and drain connections when they are no longer affine, which would be triggered by the peer reaping)